### DISCUSSION TITLE
Select a Path

### DISCUSSION THREAD

**Discussion by johnlindquist** [johnlindquist/kit#809] [Docs]
<meta sectionIndex="1">
<meta url="https://github.com/johnlindquist/kit/discussions/809">
<meta id="D_kwDOEu7MBc4AP9Tf">
<meta title="Select a Path">
<meta section="Essentials">
<meta i="2">    
<meta path="docs/select-a-path">

## Select a Path

```js
// Name: Select a Path

import "@johnlindquist/kit"

let filePath = await path()

await div(md(`You selected ${filePath}`))
```

[Open select-a-path in Script Kit](https://scriptkit.com/api/new?name=select-a-path&url=https://gist.githubusercontent.com/johnlindquist/68ae880d76f6d92b1aa9994501465f2b/raw/839a08ef025a07e5d5e292c8730d7c631b934798/select-a-path.js")

## Select a Path with Options

```js
// Name: Select a Path with Options

import "@johnlindquist/kit"

await path({
  hint: `Select a path containing JS files`,
  onlyDirs: true,
  onChoiceFocus: async (input, { focused }) => {
    let focusedPath = focused.value
    try {
      let files = await readdir(focusedPath)
      let hasJS = files.find(f => f.endsWith(".js"))

      setPreview(
        md(
          `${
            hasJS ? "✅ Found" : "🔴 Didn't find"
          } JS files`
        )
      )
    } catch (error) {
      log(error)
    }
  },
})
```

[Open select-a-path-with-options in Script Kit](https://scriptkit.com/api/new?name=select-a-path-with-options&url=https://gist.githubusercontent.com/johnlindquist/8ec7f7178cd44481aed4e968fd83da3f/raw/c8a4d8deacaa192f07fbdfed7c17a75558bc99a7/select-a-path-with-options.js")

## Drag and Drop

```js
// Name: Drop Example

import "@johnlindquist/kit"

// Note: Dropping one or more files returns an array of file information
// Dropping text or an image from the browser returns a string
let fileInfos = await drop()

let filePaths = fileInfos.map(f => f.path).join(",")

await div(md(filePaths))
```

[Open drop-example in Script Kit](https://scriptkit.com/api/new?name=drop-example&url=https://gist.githubusercontent.com/johnlindquist/f7937ef8b3d5827b5aaa17b59dc4e223/raw/183d7cdd3c3e687cdd4fd6fd833abd957e57d3de/drop-example.js")

## Select from Finder Prompts

```js
// Name: Select from Finder Prompt

import "@johnlindquist/kit"

let filePath = await selectFile()

let folderPath = await selectFolder()

await div(md(`You selected ${filePath} and ${folderPath}`))
```

[Open select-from-finder-prompt in Script Kit](https://scriptkit.com/api/new?name=select-from-finder-prompt&url=https://gist.githubusercontent.com/johnlindquist/d27e5970cb6284bd28b746eaeb49df78/raw/5f4b446f2b5a61435a651e0132e878fae9a4f819/select-from-finder-prompt.js")

**Comment by jbranchaud** [comment]
I ran into some confusion / odd-to-me behavior while trying out `path()` with an argument. It is possible that I'm not specifying the option correctly. I wanted to specify a path relative to my home directory (`~/code/til`). That didn't seem to work, so I ended up using the absolute path. Here is a video demonstrating the behavior. The format might be a little weird as I was half in the mindset of demonstrating the behavior and half in the mindset of recording a screencast.

https://user-images.githubusercontent.com/694063/202872232-bb84fa39-172f-46a9-ae10-9506d85ccc7d.mp4

**Reply by johnlindquist** [reply]
Because ~ and / are Mac/Linux specific, I recommend using the
"home" helper:

await path(home("code","til"))

I'll update the path function in the next release to support the tilde since that was your expected behavior.

Also, the fs/promises functions are global:

await readFile()

And path is global, so you can use:

path.resolve without importing anything.

**Reply by jbranchaud** [reply]
That's really helpful. didn't know about the `home()` function. It does just what I need.

A follow-up question, at timestamp 1:42 in that video above, you'll see that when I successfully open `path()` to the full path that I want, it doesn't display any files or directories for it. I have to delete and re-add the forward slash for those to start showing up. I was wondering if that was a bug. I tried looking through [the `__pathSelector` function](https://github.com/johnlindquist/kit/blob/fceebe9b61202b6da25d866a8a09699b372aed3b/src/target/app.ts#L1524) to see if I could narrow down the logic, but couldn't quite find what the issue might be.

**Reply by johnlindquist** [reply]
Definitely a bug. I'll get it sorted ASAP

**Reply by johnlindquist** [reply]
@jbranchaud if you append a final `/` to your absolute path, do you still experience the same bug? The logic for listing files in a directory is essentially "did the number of slashes change?".

**Reply by jbranchaud** [reply]
@johnlindquist I tried both of these variants (with and without `home`) and I got the same buggy behavior.

```javascript
// let tilDir = home("code", "til/");
let tilDir = "/Users/jbranchaud/code/til/";
let filepath = await path(tilDir);
```

**Reply by jbranchaud** [reply]
@johnlindquist I'm realizing this isn't the best place to be reporting this issue. If it would be helpful, I can raise this in an actual Issue.

**Reply by johnlindquist** [reply]
@jbranchaud https://github.com/johnlindquist/kit/issues/950
